### PR TITLE
Add additional doc to the method io.netty.buffer.EmptyByteBuf#isDirect

### LINKED ISSUE #11354: Counterintuitive behavior of method io.netty.buffer.AbstractByteBuf#readBytes(int)
### Expected behavior
The following code should print both `true` or both `false`:
```
public static void main(String[] args) {
        ByteBuf heapBuffer =  new PooledByteBufAllocator(false).heapBuffer(1, 1).writeByte(1);

        ByteBuf buffer_a = heapBuffer.readBytes(0);
        ByteBuf buffer_b = heapBuffer.readBytes(1);

        System.out.println(buffer_a.isDirect());
        System.out.println(buffer_b.isDirect());
    }
```

### Actual behavior
`buffer_a.isDirect()` printed `true` ,
`buffer_b.isDirect()` printed `false`.

### Steps to reproduce
See above
### Minimal yet complete reproducer code (or URL to code)
See above
### Netty version
4.1
### JVM version (e.g. `java -version`)
1.8
### OS version (e.g. `uname -a`)
MacOS

### PR DESCRIPTION
**Pull request by laosijikaichele**
Motivation:

Add additional explanation to the doc of method `io.netty.buffer.EmptyByteBuf#isDirect`, to make it more clear to understand.

Modification:

Add additional explanation to the doc of method `io.netty.buffer.EmptyByteBuf#isDirect`.

Result:

Fixes #11354.

### CODE REVIEW

File: buffer/src/main/java/io/netty/buffer/EmptyByteBuf.java [hunk 1]
```diff
@@ -105,6 +105,12 @@ public boolean isReadOnly() {
         return false;
     }
 
+    /**
+     * An empty buffer is a special buffer, it is both heap and direct,
+     * which means if there is a isHeap() method in this class, it will return {@code true} too.
+     * See <a href="https://github.com/netty/netty/issues/11354#issuecomment-855193484">this</a>
+     * for detail explanation.
+     */
```
Comment (trustin) [inline buffer/src/main/java/io/netty/buffer/EmptyByteBuf.java:113]: I guess most users will not take a look into the Javadoc of `EmptyByteBuf` because it's not part of public API. How about just adding a couple sentences like the following to `Unpooled.EMPTY_BUFFER`?

```java
    /**
     * A empty buffer whose capacity is {@code 0}. Note that an empty buffer is both direct
     * and heap buffer, because it is always backed by both a zero-length byte array and
     * direct buffer internally.
     */
    public static final ByteBuf EMPTY_BUFFER = ALLOC.buffer(0, 0);
```

Comment (laosijikaichele) [reply buffer/src/main/java/io/netty/buffer/EmptyByteBuf.java:113]: @trustin , The reason I added the doc to `EmptyByteBuf.isDirect()` is: The doc of its parent method `ByteBuf.isDirect()` states: 
>   * Returns {@code true} if and only if this buffer is backed by an
>   * NIO direct buffer.

Note the 'if and only if this buffer is backed by an NIO direct buffer' which is not a consistent logic with `EmptyByteBuf.isDirect()`, so I think it needs the additional docs to explain it.

Comment (trustin) [reply buffer/src/main/java/io/netty/buffer/EmptyByteBuf.java:113]: I see. Then how about updating the Javadoc of `ByteBuf.isDirect()`?

Comment (laosijikaichele) [reply buffer/src/main/java/io/netty/buffer/EmptyByteBuf.java:113]: I think it is OK to update the doc of `ByteBuf`, then the following methods's doc from `ByteBuf` may also need to be updated:

> ByteBuf.hasArray();
> ByteBuf.hasMemoryAddress();

because they all have doc states like "if and only if". WDYT?

Comment (trustin) [reply buffer/src/main/java/io/netty/buffer/EmptyByteBuf.java:113]: SGTM. Please go ahead.

Comment (trustin) [reply buffer/src/main/java/io/netty/buffer/EmptyByteBuf.java:113]: /cc @chrisvest You might want to update the Javadoc of the new API in your PR likewise.

Comment (chrisvest) [reply buffer/src/main/java/io/netty/buffer/EmptyByteBuf.java:113]: @trustin I think behaviour should be consistent as much as possible. In the new API the only thing special about an empty buffer is that it can't be directly allocated, but it can be created in other ways.

Comment (trustin) [reply buffer/src/main/java/io/netty/buffer/EmptyByteBuf.java:113]: @chrisvest Yeah, what I'm suggesting is to mention that an empty buffer has both array and memory address unlike other buffers. Do we already?

Comment (laosijikaichele) [reply buffer/src/main/java/io/netty/buffer/EmptyByteBuf.java]: docs updated.

Comment (chrisvest) [reply buffer/src/main/java/io/netty/buffer/EmptyByteBuf.java]: @trustin In the new API even empty buffers are either on-heap or off-heap, not some mix. There's nothing special about empty buffers in the new API, so there's nothing to document.

Comment (trustin) [reply buffer/src/main/java/io/netty/buffer/EmptyByteBuf.java]: @chrisvest I see. Thanks for letting me know!

**Review by trustin — APPROVED** [review]
Thanks, @laosijikaichele !

File: buffer/src/main/java/io/netty/buffer/ByteBuf.java [hunk 2]
```diff
@@ -307,6 +307,9 @@
     /**
      * Returns {@code true} if and only if this buffer is backed by an
      * NIO direct buffer.
+     * Note: An empty buffer is both direct and heap buffer,
+     * because it is always backed by a zero-length byte array and a direct buffer internally.
+     * See {@link EmptyByteBuf#isDirect()}
```
Comment (normanmaurer) [inline buffer/src/main/java/io/netty/buffer/ByteBuf.java:312]: @trustin @chrisvest Adding the doc here doesn't feel right as this is an implementation detail. An custom implementation of `ByteBufAllocator` / `ByteBuf` may not do this.

File: buffer/src/main/java/io/netty/buffer/ByteBuf.java [hunk 3]
```diff
@@ -2372,6 +2378,9 @@ public ByteBuf writeDoubleLE(double value) {
     /**
      * Returns {@code true} if and only if this buffer has a reference to the low-level memory address that points
      * to the backing data.
+     * Note: An empty buffer is both direct and heap buffer,
+     * because it is always backed by a zero-length byte array and a direct buffer internally.
+     * See {@link EmptyByteBuf#hasMemoryAddress()}
```
Comment (normanmaurer) [inline buffer/src/main/java/io/netty/buffer/ByteBuf.java:2383]: See above

File: buffer/src/main/java/io/netty/buffer/ByteBuf.java [hunk 4]
```diff
@@ -2349,6 +2352,9 @@ public ByteBuf writeDoubleLE(double value) {
      * Returns {@code true} if and only if this buffer has a backing byte array.
      * If this method returns true, you can safely call {@link #array()} and
      * {@link #arrayOffset()}.
+     * Note: An empty buffer is both direct and heap buffer,
+     * because it is always backed by a zero-length byte array and a direct buffer internally.
+     * See {@link EmptyByteBuf#hasArray()}
```
Comment (normanmaurer) [inline buffer/src/main/java/io/netty/buffer/ByteBuf.java:2357]: See above

File: buffer/src/main/java/io/netty/buffer/ByteBuf.java (hunk 2 above)
Comment (normanmaurer) [reply buffer/src/main/java/io/netty/buffer/ByteBuf.java:312]: Maybe we should just rephrase this to make it "optional".

Something like:

```
* Note: An empty buffer MAY be both direct and heap buffer,     
* because it is MAY be backed by a zero-length byte array and a direct buffer internally.    
* See {@link EmptyByteBuf#isDirect()}
```

Some comment true for all the others

**Review by normanmaurer — CHANGES_REQUESTED** [review]

File: buffer/src/main/java/io/netty/buffer/ByteBuf.java (hunk 2 above)
Comment (chrisvest) [reply buffer/src/main/java/io/netty/buffer/ByteBuf.java:312]: I agree with @normanmaurer here.

Comment (trustin) [reply buffer/src/main/java/io/netty/buffer/ByteBuf.java:312]: I see two options here:
- Make it a contract, so it's clear that an empty buffer is always both a heap buffer and a direct buffer; or
- Add a section in the Javadoc of `ByteBuf` that states it is possible for a `ByteBuf` to be backed by both a heap buffer and a direct buffer, mentioning `Unpooled.EMPTY_BUFFER` as an example.

Comment (normanmaurer) [reply buffer/src/main/java/io/netty/buffer/ByteBuf.java:312]: @trustin I think the second option is the best as we can't really add the contract now as it would be a "breaking change".

Comment (NiteshKant) [reply buffer/src/main/java/io/netty/buffer/ByteBuf.java:312]: +1 that the second option is better. I would also suggest to remove the slight implementation detail suggested in https://github.com/netty/netty/pull/11366/files#r646313599 and use a language like:

```
NOTE: {@link EmptyByteBuf#isDirect()} and {@link EmptyByteBuf#isHeap()} are not mutually 
exclusive, a buffer can be both heap and direct.
```

Comment (laosijikaichele) [reply buffer/src/main/java/io/netty/buffer/ByteBuf.java:312]: @trustin , @normanmaurer , @NiteshKant, @chrisvest  I think the first option is better, if we use the second option, then there will possible be a non-empty buffer with both a byte array and direct buffer, then we may should modify the logic from method like `ByteBufUtil.writeUtf8(AbstractByteBuf, int, int, CharSequence, int, int)` , because it does not cover both array and direct logic like following:
```
        if (PlatformDependent.hasUnsafe()) {
            if (buffer.hasArray()) {
                return unsafeWriteUtf8(buffer.array(), PlatformDependent.byteArrayBaseOffset(),
                                       buffer.arrayOffset() + writerIndex, seq, start, end);
            }
            if (buffer.hasMemoryAddress()) {
                return unsafeWriteUtf8(null, buffer.memoryAddress(), writerIndex, seq, start, end);
            }
        } else {
            if (buffer.hasArray()) {
                return safeArrayWriteUtf8(buffer.array(), buffer.arrayOffset() + writerIndex, seq, start, end);
            }
            if (buffer.isDirect()) {
                assert buffer.nioBufferCount() == 1;
                final ByteBuffer internalDirectBuffer = buffer.internalNioBuffer(writerIndex, reservedBytes);
                final int bufferPosition = internalDirectBuffer.position();
                return safeDirectWriteUtf8(internalDirectBuffer, bufferPosition, seq, start, end);
            }
        }
        return safeWriteUtf8(buffer, writerIndex, seq, start, end);
    }

Comment (NiteshKant) [reply buffer/src/main/java/io/netty/buffer/ByteBuf.java:312]: > then there will possible exist a non-empty buffer with both a byte array and direct buffer

Will implementing such a buffer be wrong though? If an empty buffer can have both byte array and a direct buffer is this technically wrong for an non-empty buffer to have both too? As long as both direct and heap memory has same data, the code you have referred here will work, correct? So, the question is does it help us in any way by making the contract stricter?

Comment (laosijikaichele) [reply buffer/src/main/java/io/netty/buffer/ByteBuf.java:312]: @NiteshKant As you said:

> As long as both direct and heap memory has same data, the code you have referred here will work, correct?

No it will not work. 
Think about this question: 
How to make both direct and heap memory has same data when you call the public util method like 'ByteBufUtil.writeUtf8(ByteBuf, CharSequence)'? 
The answer is you have to modify the logic of 'ByteBufUtil.writeUtf8(ByteBuf, CharSequence)', right? That is what I said "not good" because it break the origin contract "**if and only if**" in a larger scale.
